### PR TITLE
Stock movement registry

### LINKED ISSUE #3968: Registre d'acceptation
We have received this feature request:

> J'ai pensé proposer l’implémentation d'un registre dit d'acceptation, dans le module stock. En effet, le souci étant de rester positivement plus proche des pratiques d'usage dans le monde professionnel des hôpitaux, ce registre est un des documents demandés lors des missions de supervision. Il va faire référence au document généré quand on soumet une entrée en stock en provenance de l’extérieur (voir image en PJ). Seulement que la présentation du registre sera exactement comme celle du registre des achats; quand on clique sur voir le document, ça nous ramène au document en image ci haut mentionné.
Merci pour votre attention et franche collaboration.

Basically, we need to show the same level of detail that we do for a purchase order registry, but for all stock entries.  It might be a new module, or an enhancement to the existing stock entry registry.

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR is related to the issue #3968 

Current stock movement registry

image

The new stock inline movement registry

image

Close #3968

### CODE REVIEW

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

File: server/controllers/stock/reports/stock_inline_movements.report.handlebars [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            <!-- blank row  -->
+            {{#unless @last }}
+              <!-- blank line -->
+              <tr style="border:none;">
+                <th style="border:none;"></th>
+              </tr>
+            {{/unless}}
+
+          {{else}}
+            {{> emptyTable columns=6}}
```
Comment (jniles) [inline server/controllers/stock/reports/stock_inline_movements.report.handlebars]: Are you sure this is columns=6 and not columns=5?

**Review by jniles — COMMENTED** [review]
@mbayopanda I'm not sure what functionality this adds.  Can you please explain it to me?  It looks like it just repeats the old regsitry, but with less columns...

**Comment by jniles** [comment]
@mbayopanda I'm okay with these changes.  Can you please replace the old registry with this one?  I would do it by updating the unit table, the report table, and the ui-router routes.  That way, it will never appear side by side with the other registry.

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

**Comment by mbayopanda** [comment]
@jniles changes landed

File: client/src/modules/stock/stock.routes.js [hunk 2]
```diff
@@ -10,14 +10,14 @@ angular.module('bhima.routes')
...
+    // .state('stockMovements', {
+    //   url         : '/stock/movements',
+    //   controller  : 'StockMovementsController as StockCtrl',
+    //   templateUrl : 'modules/stock/movements/registry.html',
+    //   params : {
+    //     filters : [],
+    //   },
+    // })
 
       .state('stockInlineMovements', {
```
Comment (jniles) [inline client/src/modules/stock/stock.routes.js]: Can you replace this `stockInlineMovements` with `stockMovements`?  We have a lot of links from other registries to the `stockMovements` (Inventory, Purchase Orders, etc) and it would be nice to ensure those still worked.

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

**Comment by jniles** [comment]
Looks like the tests need to be updated for this to properly work.

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

File: client/src/modules/stock/movements/modals/search.modal.html [hunk 3]
```diff
@@ -19,7 +19,6 @@
           <div class="form-group">
             <bh-clear on-clear="$ctrl.clear('is_exit')"></bh-clear>
             <br/>
-            <!-- Gender -->
```
Comment (jniles) [inline client/src/modules/stock/movements/modals/search.modal.html:22]: 😆

File: server/controllers/stock/core.js [hunk 4]
```diff
@@ -298,6 +299,42 @@ async function getLotsMovements(depotUuid, params) {
...
+    d.text AS depot_text, IF(is_exit = 1, "OUT", "IN") AS io,
+    BUID(m.depot_uuid) AS depot_uuid, m.is_exit, m.date, BUID(m.document_uuid) AS document_uuid,
+    m.flux_id, BUID(m.entity_uuid) AS entity_uuid, SUM(m.unit_cost * m.quantity) AS cost,
+    f.label AS flux_label, BUID(m.invoice_uuid) AS invoice_uuid, dm.text AS documentReference
+  FROM stock_movement m
+  JOIN lot l ON l.uuid = m.lot_uuid
+  JOIN depot d ON d.uuid = m.depot_uuid
+  JOIN flux f ON f.id = m.flux_id
+  LEFT JOIN document_map dm ON dm.uuid = m.document_uuid
+  LEFT JOIN service AS serv ON serv.uuid = m.entity_uuid
```
Comment (jniles) [inline server/controllers/stock/core.js:330]: Can we JOIN this on inventory?  It looks like this breaks the link from articles in stock.  See below:

wdqCZwmxK7

File: test/end-to-end/stock/stock.movements.spec.js [hunk 5]
```diff
@@ -26,66 +26,52 @@ function StockMovementsRegistryTests() {
...
+    await GU.expectRowCount(gridId, 2 + depotGroupingRow);
   });
 
   it('find movements by Service', async () => {
     await modal.setServiceUuid('Administration');
     await modal.submit();
-    await GU.expectRowCount(gridId, 2);
-  });
-
-  it('find movements by inventory', async () => {
```
Comment (jniles) [inline test/end-to-end/stock/stock.movements.spec.js:66]: Why did you remove this functionality?

File: server/models/migrations/next/migrate.sql [hunk 6]
```diff
@@ -1,3 +1,36 @@
 /* migration script from the version 1.13.1 to the next one */
 
 ALTER TABLE `unit` DROP COLUMN `url`;
+-- migrate next.sql
+
+/**
+@author: jniles:
+@date: 2020-05-14
+*/
+UPDATE `report` SET `report_key` = 'stock_sheet' WHERE `title_key` = 'REPORT.STOCK.INVENTORY_REPORT';
```
Comment (jniles) [inline server/models/migrations/next/migrate.sql]: I think a lot of these are from a previous migration script.  Can you remove them?

File: server/controllers/stock/reports/stock_inline_movements.report.handlebars [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+          {{/each}}
+        </tbody>
+        <tfoot>
+          <tr style="font-weight: bold;">
+            <td colspan="5">{{rows.length}} {{translate 'STOCK.ITEMS'}}</td>
```
Comment (jniles) [inline server/controllers/stock/reports/stock_inline_movements.report.handlebars:79]: It looks like this might not be translated.

image

Also, does it make sense to have these in both locations?  I can see arguments for both, but am undecided myself.

File: server/controllers/stock/reports/stock/inline_movements_report.js [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+  let display = {};
+  const optionReport = _.extend(req.query, pdfOptions, {
+    filename : 'TREE.STOCK_INLINE_MOVEMENTS',
+    csvKey : 'rows',
+    renameKeys : false,
+  });
+
+  // set up the report with report manager
+  try {
+    if (req.query.displayNames) {
```
Comment (jniles) [inline server/controllers/stock/reports/stock/inline_movements_report.js]: I think the point of this is to try and put the display names in the report, but that doesn't work.  See below:

image


Personally, I prefer using the method of [parseDisplayValue()](https://github.com/IMA-WorldHealth/bhima/blob/master/server/controllers/finance/reports/shared.js#L208) which is used in the `reports/shared.js` version of `formatFilters()`.

Comment (jniles) [reply server/controllers/stock/reports/stock/inline_movements_report.js]: You can either try to make this work, or just remove the `try`/`catch` and we'll update it later.  I think many reports suffer the same problem because developers copy and pasted code without understanding what it needed to work.

File: server/controllers/stock/reports/stock/inline_movements_report.js [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+      display,
+      filters : formatFilters(req.query),
+    };
+
+    // group by depot
+    let depots = _.groupBy(rows, d => d.depot_text);
+
+    // make sure that they keys are sorted in alphabetical order
+    depots = _.mapValues(depots, lines => {
+      _.sortBy(lines, 'depot_text');
```
Comment (jniles) [inline server/controllers/stock/reports/stock/inline_movements_report.js]: Did you know you can actually chain all these?  I learned about it recently.  [Here is an example](https://github.com/IMA-WorldHealth/bhima/blob/master/server/controllers/inventory/reports/changes.js#L91).

I think it should be something like:
```js
const depots = _.chain(rows)
  .groupBy(d => d.depot_text)
  .mapValues(lines => _.sortBy(lines, 'depot_text'))
  .value();
```

See if you can get something like that to work.

**Review by jniles — COMMENTED** [review]
@mbayopanda, thanks for these changes!  I've noticed a few things that don't work and made some comments on them.

File: test/end-to-end/stock/stock.movements.spec.js (hunk 5 above)
Comment (mbayopanda) [reply test/end-to-end/stock/stock.movements.spec.js:66]: There was not any join with the inventory table so i was thinking that it was not necessary to have this test, since the registry was focused on just movement (stock_movement table) information

File: server/controllers/stock/reports/stock/inline_movements_report.js (hunk 8 above)
Comment (mbayopanda) [reply server/controllers/stock/reports/stock/inline_movements_report.js]: Why remove the `try/catch` ? There is `await Stock.getMovements(null, req.query);` and `await report.render(data);` which may thrown errors

File: server/controllers/stock/reports/stock_inline_movements.report.handlebars (hunk 7 above)
Comment (mbayopanda) [reply server/controllers/stock/reports/stock_inline_movements.report.handlebars:79]: image

In case of multiple depots, value in the line of the depot name is the total of the depot and the value in the bottom is the total of items loaded for all depots

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

File: server/controllers/stock/reports/stock/inline_movements_report.js (hunk 8 above)
Comment (jniles) [reply server/controllers/stock/reports/stock/inline_movements_report.js]: Sorry, the `try`/`catch` around the `req.query.displayNames`.  Basically, remove the entire block, since it isn't functional and a future developer shouldn't think that code does something important if it doesn't do anything at all.

File: server/controllers/stock/reports/stock_inline_movements.report.handlebars (hunk 7 above)
Comment (jniles) [reply server/controllers/stock/reports/stock_inline_movements.report.handlebars:79]: I agree.

**Comment by kwilu** [comment]
Hello @mbayopanda, it seems like you have referenced #3968 in your pull request description, but you have not referenced them in your commit message description(s). When you reference an issue in a commit message, it automatically closes the corresponding issue when the commit is merged.

Please run `git commit --amend` in your command line client to amend your commit message description with `Fixes #3968.`.

An example of a correctly-formatted commit:
```
commit fabd5e450374c8dde65ec35f02140383940fe146
Author: kwilu
Date:   Sat Mar 18 13:42:40 2017 -0700

    pull requests: Check PR commits reference when issue is referenced.

    Fixes #51.
```

Thank you for your contributions to BHIMA!

File: client/src/modules/stock/movements/registry.js [hunk 10]
```diff
@@ -233,20 +188,13 @@ function StockMovementsController(
...
       .catch(Notify.handleError)
       .finally(toggleLoading);
   }
 
-  function orderByDepot(rowA, rowB) {
-    return rowA.depot_text > rowB.depot_text ? 1 : -1;
-  }
-
   function handleMovementRows(rows) {
-    // FIXME(@jniles): we should do this ordering on the server via an ORDER BY
```
Comment (jniles) [inline client/src/modules/stock/movements/registry.js:247]: 👍

**Review by jniles — APPROVED** [review]
Code LGTM.

bors r+
